File: src/main/java/com/mallang/auth/application/OauthService.java
package com.mallang.auth.application;

import com.mallang.auth.domain.OauthId.OauthServerType;
import com.mallang.auth.domain.OauthMember;
import com.mallang.auth.domain.OauthMemberRepository;
import com.mallang.auth.domain.oauth.AuthCodeRequestUrlProviderComposite;
import com.mallang.auth.domain.oauth.OauthMemberClientComposite;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@RequiredArgsConstructor
@Service
public class OauthService {

    private final OauthMemberRepository oauthMemberRepository;
    private final OauthMemberClientComposite oauthMemberClientComposite;
    private final AuthCodeRequestUrlProviderComposite authCodeRequestUrlProviderComposite;

    public String getAuthCodeRequestUrl(OauthServerType oauthServerType) {
        return authCodeRequestUrlProviderComposite.provide(oauthServerType);
    }

    public Long login(OauthServerType oauthServerType, String authCode) {
        OauthMember member = oauthMemberClientComposite.fetch(oauthServerType, authCode);
        return oauthMemberRepository.findByOauthId(member.getOauthId())
                .orElseGet(() -> signup(member))
                .getId();
    }

    private OauthMember signup(OauthMember member) {
        return oauthMemberRepository.save(member);
    }
}

File: src/main/kotlin/com/mallang/auth/application/OauthService.kt
package com.mallang.auth.application

import com.mallang.auth.domain.Member
import com.mallang.auth.domain.MemberRepository
import com.mallang.auth.domain.OauthServerType
import com.mallang.auth.domain.oauth.AuthCodeRequestUrlProviderComposite
import com.mallang.auth.domain.oauth.OauthMemberClientComposite
import org.springframework.stereotype.Service

@Service
class OauthService(
        private val memberRepository: MemberRepository,
        private val oauthMemberClientComposite: OauthMemberClientComposite,
        private val authCodeRequestUrlProviderComposite: AuthCodeRequestUrlProviderComposite
) {

    fun getAuthCodeRequestUrl(oauthServerType: OauthServerType): String {
        return authCodeRequestUrlProviderComposite.provide(oauthServerType)
    }

    fun login(oauthServerType: OauthServerType, authCode: String): Long {
        val member = oauthMemberClientComposite.fetch(oauthServerType, authCode)
        return memberRepository.findByOauthId(member.oauthId)
                .orElseGet { signUp(member) }
                .id
    }

    private fun signUp(member: Member): Member {
        return memberRepository.save(member)
    }
}
